pkg/provider/doppler/auth_oidc.go

```diff
@@ -0,0 +1,245 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package doppler
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	esv1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1"
+)
+
+const (
+	defaultTokenTTL = 600
+	minTokenBuffer  = 60
+	dopplerOIDCPath = "/v3/auth/oidc"
+)
+
+type OIDCTokenManager struct {
+	corev1    typedcorev1.CoreV1Interface
+	store     *esv1.DopplerProvider
+	namespace string
+	storeKind string
+	storeName string
+	baseURL   string
+	verifyTLS bool
+
+	mu          sync.RWMutex
+	cachedToken string
+	tokenExpiry time.Time
+}
+
+func NewOIDCTokenManager(
+	corev1 typedcorev1.CoreV1Interface,
+	store *esv1.DopplerProvider,
+	namespace string,
+	storeKind string,
+	storeName string,
+) *OIDCTokenManager {
+	baseURL := "https://api.doppler.com"
+	if customURL := os.Getenv(customBaseURLEnvVar); customURL != "" {
+		baseURL = customURL
+	}
+
+	verifyTLS := true
+	if customVerifyTLS := os.Getenv(verifyTLSOverrideEnvVar); customVerifyTLS == "false" {
+		verifyTLS = false
+	}
+
+	return &OIDCTokenManager{
+		corev1:    corev1,
+		store:     store,
+		namespace: namespace,
+		storeKind: storeKind,
+		storeName: storeName,
+		baseURL:   baseURL,
+		verifyTLS: verifyTLS,
+	}
+}
+
+func (m *OIDCTokenManager) GetToken(ctx context.Context) (string, error) {
+	m.mu.RLock()
+	if m.isTokenValid() {
+		token := m.cachedToken
+		m.mu.RUnlock()
+		return token, nil
+	}
+	m.mu.RUnlock()
+
+	return m.refreshToken(ctx)
+}
+
+func (m *OIDCTokenManager) isTokenValid() bool {
+	if m.cachedToken == "" {
+		return false
+	}
+	return time.Until(m.tokenExpiry) > minTokenBuffer*time.Second
+}
+
+func (m *OIDCTokenManager) refreshToken(ctx context.Context) (string, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.isTokenValid() {
+		return m.cachedToken, nil
+	}
+
+	saToken, err := m.createServiceAccountToken(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to create service account token: %w", err)
+	}
+
+	dopplerToken, expiry, err := m.exchangeTokenWithDoppler(ctx, saToken)
+	if err != nil {
+		return "", fmt.Errorf("failed to exchange token with Doppler: %w", err)
+	}
+
+	m.cachedToken = dopplerToken
+	m.tokenExpiry = expiry
+
+	return dopplerToken, nil
+}
+
+func (m *OIDCTokenManager) createServiceAccountToken(ctx context.Context) (string, error) {
+	oidcAuth := m.store.Auth.OIDCConfig
+
+	audiences := []string{m.baseURL}
+
+	// Add resource-specific audience for cryptographic binding
+	if m.storeKind == esv1.ClusterSecretStoreKind {
+		audiences = append(audiences, fmt.Sprintf("clusterSecretStore:%s", m.storeName))
+	} else {
+		audiences = append(audiences, fmt.Sprintf("secretStore:%s:%s", m.namespace, m.storeName))
+	}
```

Comment:
nit/blocking: The type used for `serviceAccountRef` in the CRD includes an `audiences` property and the docs indicate these will be included in the resulting JWT, but they aren't being passed through.